### PR TITLE
Forward port latest build system from node-cpp-skel

### PR DESCRIPTION
**Pull request by springmeyer**
This brings in the latest build scaffolding from [node-cpp-skel](https://github.com/mapbox/node-cpp-skel), effectively modernizing dawg-cache.

This solves #5, #6, #16, #20, #21 and now provides:

 - A welcoming Makefile that allows easy source compiles with`make`
 - Run `clang-tidy` checks with `make tidy`
 - Run `clang-format` checks with `make format`
 - Testing of address, undefined, and leak sanitizer on travis
 - Support for multiple build types for published binaries: Release, vs Debug, vs Sanitized

More details on the build setup at https://github.com/mapbox/node-cpp-skel/blob/master/docs/extended-tour.md#builds

### CODE REVIEW

**Comment by springmeyer** [comment]
This is ready except that the clang-format and clang-tidy jobs on travis are failing because:

  - Clang-format changes need to be applied to the code
  - There are a lot of issues that clang-tidy errors on in the code, some probably very minor, others potential bugs.

My plan is to:

 - Create a new branch and PR on top of this one that applies the clang-format fixes
 - Create a new branch and PR on top of the clang-format fixes branch that starts making fixes to address the clang-tidy errors

The motivation of ^^ approach is to provide PRs that make the changes easier to reason about on their own. Once the above PRs look good, I'll plan to merge into this once, before merging this one into master.

**Comment by apendleton** [comment]
@springmeyer happy to defer to your judgment here with respect to the build scripts and such. The old ones I'm sure I copied from some existing repo at the time (carmen-cache I bet?) without really closely examining what they did.

As for the code changes: looks to me like a couple of style tweaks which look fine, and changes from reinterpret_cast to memcpy in a couple of other places. The former look good to me, and the latter we've already done (for the same reason) in carmen-cache, so also good.

**Review by apendleton — APPROVED** [review]

**Comment by springmeyer** [comment]
👍 Thanks for the review @apendleton. I've just merged the clang-tidy/format fixes into this branch and once this is green I'll merge this into master. That will set us up to hit these issues next:

 - https://github.com/mapbox/dawg-cache/issues/25
 - https://github.com/mapbox/dawg-cache/issues/26

**Pull request merged**
